### DISCUSSION TITLE
WYSIWYG "Add/Edit Link" function allows linking to directus files

### DISCUSSION THREAD

**Discussion by bastian-41** [directus/directus#19659] [Feature Requests]
### Summary

The "Add/Edit Link" feature in the WYSIWIG editor should allow me to optionally choose (or upload) a file stored in directus.

### Motivation

When using Directus in an CMS like scenario, users upload files (e.g. pdf documents) into directus and then want to link them in a free text provided by the WYSIWIG editor. The link should be rendered as an <a> tag.  While it IS possible to enter the URL manually, it is cumbersome and error prone.

The features "Add/Edit Image" and "Add/Edit Media" already support a similar feature. But only as an <img> oder <video> tag. Not as an <a> tag.

### Detailed Design

When clicking on "Add/Edit Link" in the WYSIWYG editor, first a pop up appears, where I can choose to upload a file, choose a file from the library, or enter the URL manually. After choosing one option, I'll reach the existing "Add/Edit Link" Modal, and the URL field is prefilled (according to the option used beforehand).
The existing features "Add/Edit Image" and "Add/edit Media" work similarily.

### Must Have:
- Add/Edit Link allows me to choose a file from the directus files.

### Could Have:
- Two different items in the WYSIWIG Toolbar "Add/Edit Link" (working as it is now and "Add/Edit Link (to Directus File)" working as described above.

### Drawbacks

- Since the feature requests relates to an existing feature it might confuse users, if the behavior changes (therefore the "could have" requirement which moves the funcionality into a new Item/Button which a directus developer/designer must choose explicitly).

### Alternatives

I did try out to use the existing "Add/Edit Media" or "Add/Edit Image" functions and fixing the created html markup on the fly when exporting via Directus REST API. Basically replaying img or video tags by a tags.

But this is

1. Not elegant at all and
2. an a-Tag needs its an Display Text, which the video and image feature doesn't have / support

### Adoption Strategy

When this new feature is optionally implemented (e.g. a new Toolbox item "Add/Edit Link to Directus file", it won't affect existing solutions. No migration is needed.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by bastian-41** [reply]
Hey @rijkvanzanten, is a review meeting already scheduled? I've several demands which might profit from this feature request. But since I cannot postpone the demands indefinitely I'd like to have a hint if this feature request might be granted in the foreseeable future or if I should think about other solutions outside of Directus. My preference is clear. 😁

**Reply by rijkvanzanten** [reply]
We discuss feature requests every two weeks on Discord 👍🏻 This one hasn't come up yet, as we go top to bottom based on upvotes

**Comment by w0kyj** [comment]
Request Review Notes:
- File permissions is a big consideration
    - Easy debug permissions because users can see images/videos
    - Need a way to set/inherit/preview permissions to files
        - inherit permissions of the current object?
        - assets endpoint doesn't know anything about the current item
            - add a way for files to use or have a unique link related to the current item
            - 
    - Difficult as Directus doesn't know the end use case / API access
    - Nested relational permissions (another discussion --> that discusses this)
- Security vs Usability
    - When adding a file link, must ensure security implications understood, but also needs to be easy for content editors to use
    - Shouldn't use access tokens as that is a security issue and is visible externally
    - Use signed JWT scoping the access to the current Item
        - How does this work with SSR, API builds that use the permissions, but now it is viewed by external unauthenticated source?
- Relative link options
    - Currently is full url and is an issue during migrations 
    - 
- Custom web components
    - Similar to Nuxt integrations
- Short Code Options
    - Stored and rendered automatically
- Use a block editor
    - Instead of supporting WYSIWYG at all -- move to block editor
        - Add a new endpoint parameter that will auto render the json to html

FEATURES TO CONSIDER:
- need a general file menu item to insert a file as a-tag
    - add to all rich text interfaces (markdown & html) already supported for block editor
- Consider Block Editor update:
    - If Editor.js supports video preview add this a a block option to match current WYSIWYG feature
    - Note markdown doesn't support so not needed

Please add any additional requirements if I missed anything :)

**Reply by nicksergeant** [reply]
Awesome to see some movement on this issue. This has come up in two other closed discussions as well:

- https://github.com/directus/directus/discussions/6605
- https://github.com/directus/directus/discussions/16505

So this could be high-impact. For me not being able to select-and-upload-files in WYSIWYG editor is the only reason we can't support certain types of clients on Directus and we have to use a different system for those clients.

> Instead of supporting WYSIWYG at all -- move to block editor

I realize you just had the review meeting but was this a final decision? Our clients desperately need the ability to select files to link content within the WYSIWYG editor, sadly.

**Reply by rijkvanzanten** [reply]
Oh no not at all, that was just a brainwave 🙂 The recording for this session will be up on https://directus.io/tv soon

**Reply by nicksergeant** [reply]
Awesome!

**Reply by rijkvanzanten** [reply]
Also the final decision here is that the MVP described above can be implemented (PRs very welcome!) and that we want to circle back later to the topics outlined below, as there's a lot of quality of life features around this topic

**Reply by nicksergeant** [reply]
@rijkvanzanten Hi! Just checking in on this feature request again. We have a town/government client who would hugely benefit from this. Is there a way we can sponsor this feature development with $$?

**Reply by rijkvanzanten** [reply]
Hi @nicksergeant! Just to make sure I don't end up implementing something half-done; are you specifically talking about the ability to select a bit of text, hit a button, choose a file, and have that bit of text link to the file you chose? Or does your ideal implementation look a bit differently? Part of the challenge with all feature requests is to make sure everybody is on the same page with what the ideal state is 🙂

**Reply by nicksergeant** [reply]
@rijkvanzanten that feature is certainly a nice to have but even just a very simple flow of:

1. Click a document icon in the WYSIWYG editor
2. Opens a very similar modal like the image modal:

![image](https://github.com/user-attachments/assets/afcf39eb-9486-4885-8503-c87f455afa74)

3. You can upload or select any file in the File Library
4. Selecting a file inserts the asset URL into the WYSIWYG editor

This solves the primary pain of not being able to easily find the public link to a file in the File Library for use in the WYSIWYG editor when you are editing page content.

Right now the work around is that admins have to manually go into the File Library, find the file, click "File Details" in the sidebar, right click on "Open in New Window", and "Copy Link Address". This is borderline impossible for our non-tech admins 😅.

There was a workaround using the Add/Edit Image flow noted [here](https://github.com/directus/directus/discussions/6605#discussioncomment-5457045), however we have noticed that this workaround doesn't work for PDF files. If you attempt to select a PDF file, it actually throws a WYSIWYG editor error in the UI (and most of the time, our government clients are trying to select PDFs to link to in their body text):

https://github.com/user-attachments/assets/955c5165-c5fb-40df-b0e7-391b12fdc3a3

**Reply by kbkmn** [reply]
@rijkvanzanten that is exactly what i need also. I have an elderly client and jumping hoops with getting link through image field and pasting it into link field is proving to be hard. Even the simplest, basic solution will me much appreciated

**Comment by bastian-41** [comment]
I'm astonished that finally there seems to be some movement regarding my old feature request.
For the moment I can't / won't contribute anything. We've solved our use case in another way. Meaning without Directus.

But I'm still convinced this feature would be a nice one.

**Comment by qntmplation** [comment]
Just jumping in to say I have a client with a catalog site that has ~1000 products manuals. Without the ability to _easily_ inline PDF document links, the site would be unmaintainable for their content editors, so Directus may not be an option for them simply for this reason. I'm not sure its something I could develop solo, but if there is any community work done on this so far, I'd love to contribute (maybe I am just not seeing it).

**Reply by w0kyj** [reply]
I think this will provide what you are looking for. Check out this directus-labs plugin:
https://github.com/directus-labs/extensions/tree/main/packages/input-rich-text-html-file-link

The Core Team with community help creates these extensions:
https://github.com/directus-labs/extensions

**Reply by qntmplation** [reply]
Nice, this does function, although it is pretty rough around the edges. Thanks for sharing! I will make further feature requests there.
